### PR TITLE
feat: add per-request key document

### PR DESCRIPTION
**Pull request by sr-remsha**

### CODE REVIEW

File: docs/Roles and Access Control/3.API Keys.md [hunk 1]
```diff
@@ -33,3 +35,39 @@ To create and configure access control for API keys:
...
+2. AI DIAL Core stores an additional data associated with the per-request key either in memory or in Redis. The data may include but is not limited to: JWT of the user or API key of the application, `trace-id`, `core-parent-span-id`, list of the user files attached to the request.
+3. AI DIAL Core passes the per-request key to the application or model adapter in the API-key header. The application or model adapter then send the per-request key back to AI DIAL Core using the same API-key header for model requests or file storage API requests.
+4. AI DIAL Core invalidates the per-request key when the request is complete.
+
+### Calculating Costs & Limits
+
+Per-request keys can be used to attribute costs and limits incurred by a request to a model to the user or application that initiated the request.
+
+To attribute costs and limits to the request originator, `trace-id` should be included in the request header. Incorrect `trace-id` will fail the request with the error 401 unauthorized. `trace-id` is saved throughout the lifespan of the request in memory or in Redis.
```
Comment (astsiapanay) [inline docs/Roles and Access Control/3.API Keys.md:56]: remove the sentence `Incorrect `trace-id` will fail the request with the error 401 unauthorized.`

Comment (astsiapanay) [inline docs/Roles and Access Control/3.API Keys.md:56]: change `To attribute costs and limits to the request originator, trace-id should be included in the request header.` -> `To attribute costs and limits to the request originator, `trace parent` should be included in the request header.`

**Review by astsiapanay — CHANGES_REQUESTED** [review]

File: docs/Roles and Access Control/3.API Keys.md (hunk 1 above)
Comment (sr-remsha) [reply docs/Roles and Access Control/3.API Keys.md]: removed

Comment (sr-remsha) [reply docs/Roles and Access Control/3.API Keys.md]: changed

File: docs/Roles and Access Control/3.API Keys.md [hunk 2]
```diff
@@ -33,3 +35,51 @@ To create and configure access control for API keys:
...
+3. AI DIAL Core passes this new outgoing per-request key in the request header. The application or model adapter then send the per-request key back to AI DIAL Core using the same API-key header for model requests or file storage API requests.
+4. AI DIAL Core invalidates the outgoing per-request key when the request is completed.
+
+### Calculating Costs & Limits
+
+Per-request keys can be used to attribute costs and limits incurred by a request to a model to the user or application that initiated the request.
+
+To attribute costs and limits to the request originator, `traceparent` should be included in the request header. `trace-id` is saved throughout the lifespan of the request in memory or in Redis.
+
+The application can as well propagate a `traceparent` header for the tracing feature. If the value of the `trace-id` propagated in the `traceparent` header does not match with the `trace-id` stored in the additional data associated with the per-request key, the application will receive 400 Bad request error.
```
Comment (astsiapanay) [inline docs/Roles and Access Control/3.API Keys.md:70]: Please remove the entire paragraph. It's not relevant any longer.

File: docs/Roles and Access Control/3.API Keys.md [hunk 3]
```diff
@@ -33,3 +35,51 @@ To create and configure access control for API keys:
...
+
+### Gathering Statistics
+
+For calculating statistics, AI DIAL Core uses `trace-id` and `core-parent-span-id` that are stored in the additional data linked to the per-request key.
+
+### Files Sharing
```
Comment (astsiapanay) [inline docs/Roles and Access Control/3.API Keys.md:76]: Please add sections for auto sharing input and output files.
There is a vital info missed:
```
To provide the path of the folder for output files, we suggest to extent the [GET /v1/bucket](https://epam-rail.com/dial_api#/paths/~1v1~1bucket/get) response with an additional field - appdata field.
If the [GET /v1/bucket](https://epam-rail.com/dial_api#/paths/~1v1~1bucket/get) request is done with a Per-request key the response will contain appdata field with the path to the shared application folder, including the User's bucket and the correct deployment-id for the application as registered in the Core configs.

GET /v1/bucket
{
  "bucket": "{application-bucket-id}",
  "appdata": "{user-bucket-id}/appdata/{deployment-id}",
}
```

**Review by astsiapanay — CHANGES_REQUESTED** [review]

File: docs/Roles and Access Control/3.API Keys.md (hunk 2 above)
Comment (sr-remsha) [reply docs/Roles and Access Control/3.API Keys.md]: removed

File: docs/Roles and Access Control/3.API Keys.md (hunk 3 above)
Comment (sr-remsha) [reply docs/Roles and Access Control/3.API Keys.md:74]: added

**Review by astsiapanay — APPROVED** [review]

**Pull request merged**